File: problems-extend-adapt/3401-3500/3424_rearrange-or-retune/solution.py
from typing import List, Optional


class Solution:
    def minRetuneCost(self, arr: List[int], brr: List[int], k: int) -> int:
        # Splitting into singleton blocks already realizes any permutation,
        # so one paid rearrangement is all Operation 1 can offer; matching
        # sorted to sorted then minimizes sum |a - b| over permutations.
        # The answer is the cheaper of leaving arr put and paying k plus
        # that matched cost. Costs reach 4 * 10^10, past 32-bit range.
        direct = 0
        for a, b in zip(arr, brr):
            d = a - b
            direct += d if d >= 0 else -d
        matched = k
        for a, b in zip(sorted(arr), sorted(brr)):
            d = a - b
            matched += d if d >= 0 else -d
        return min(direct, matched)

File: problems-extend-adapt/3401-3500/3424_rearrange-or-retune/solution.rs
impl Solution {
    pub fn min_retune_cost(arr: Vec<i32>, brr: Vec<i32>, k: i64) -> i64 {
        // Splitting into singleton blocks already realizes any
        // permutation, so one paid rearrangement is all Operation 1 can
        // offer; matching sorted to sorted then minimizes sum |a - b|
        // over permutations. The answer is the cheaper of leaving arr put
        // and paying k plus that matched cost. Sums reach 2 * 10^10 and
        // k itself is up to 2 * 10^10, so everything widens to i64.
        let direct: i64 = arr.iter().zip(&brr).map(|(&a, &b)| (a as i64 - b as i64).abs()).sum();
        let mut sa = arr.clone();
        let mut sb = brr.clone();
        sa.sort();
        sb.sort();
        let matched = k + sa
            .iter()
            .zip(&sb)
            .map(|(&a, &b)| (a as i64 - b as i64).abs())
            .sum::<i64>();
        direct.min(matched)
    }
}
